ipld/hamt/fuzz/corpus/extensions

```diff
@@ -0,0 +1 @@
+../../../../testing/fuzz-corpora/corpus/hamt_simple
```

Comment:
I would suggest a separate corpus directory. The corpus optimization and compaction is done based on coverage information, I think the two fuzzing targets will remove corpus elements which are not exercising code path useful to them.

Reply:
I would suggest creating a new directory in the submodule with a copy of the hamt_simple corpus, you can push straight to master there IMO (CI automation does that when it finds new valuable test cases).

> Every time I point cargo at a new ref-fvm branch on Github from the builtin-actors repo it takes ages to clone another copy of the fuzz targets submodule

Do you mean cloning the fuzz-corpora takes a long time? If so we could probably discard old commits there.

Looking at the CI, the corpus updating broke at some point, I'm not sure why. I will open an issue.


> Isn't there a way to just fuzz on random input for a limited time and only print failures on CI?

There isn't much point, fuzzing takes a while and works best with good corpus. One of the CID bugs was found after weeks of fuzzing on the CI.

Reply:
> Do you mean cloning the fuzz-corpora takes a long time?

The fuzz corpora and some other test vectors. For example the following shows how long it takes to initialize these two for a branch of `builtin-actors` which is built against a specific branch of `ref-fvm`:

```console
❯ cargo cache -a                                                                                                                                                                                                 ─╯
Clearing cache...

... just to make sure I don't have it ...
❯ make all                                                                                                                                                                                                     ─╯
cargo test --test storage_footprint
    Updating git submodule `https://github.com/filecoin-project/fvm-test-vectors`
    Updating git submodule `https://github.com/filecoin-project/ref-fvm-fuzz-corpora.git`
   Compiling fvm_ipld_blockstore v0.1.1 (https://github.com/filecoin-project/ref-fvm?branch=evm-891-hamt-link-extension#7fe44ee3)
   
^Cmake: *** [Makefile:31: measure-storage-footprint] Interruptuntime                                                                                                                                             

    ~/projects/consensuslab/builtin-actors/actors/evm   evm-891-hamt-link-extens… ················································································ ✘ INT   2m 32s   1.63.0  16:49:09 ─╮
❯ cargo cache -a                                                                                                                                                                                                 ─╯
Clearing cache...

Cargo cache '/home/aakoshh/.cargo':

Total:                                      1.97 GB => 1.40 GB
  19 installed binaries:                             469.61 MB
  Registry:                             784.14 MB => 710.38 MB
    Registry index:                                  358.90 MB
    2757 crate archives:                             351.47 MB
    272 => 0 crate source checkouts:          73.77 MB => 0  B
  Git db:                               713.76 MB => 224.85 MB
    32 bare git repos:                               224.85 MB
    2 => 0 git repo checkouts:               488.91 MB => 0  B

Size changed 1.97 GB => 1.40 GB (-562.67 MB, -28.59%)

    ~/projects/consensuslab/builtin-actors/actors/evm   evm-891-hamt-link-extens… ···································································································  1.63.0  16:49:45 ─╮
❯                                                                                                                                                                                                                ─╯
```

So it took 2.5 minutes to clone those two, and after that clearing it away shows it was 560MB that had to be downloaded.
By my count it took about 1 minute to do the fuzz corpora, although looking at the data on disk, it doesn't seem to be that big actually. 

Anyway, happy to do the copy, no problem, thanks for explaining how it works!

Reply:
> Interesting, I need to learn more about how this fuzzing library works.

It is less about the lib itself (as it is a wrapper/entry point for libfuzzer) but tooling around it. We use ClusterFuzzLite for fuzzing on the CI. libfuzzer is able to do coverage guided fuzzing, the corpus minimization is also feature from it but you have call it separately which ClusterFuzzLite does.

Reply:
Thanks! 
As a heads up I also added a fuzz target in https://github.com/filecoin-project/ref-fvm/pull/946

Reply:
Heads up on possible bug in CI fuzzing: https://github.com/filecoin-project/ref-fvm/issues/959

Reply:
Can't see it yet, but I'll update that PR too, thanks.